lib/active_merchant/billing/gateways/checkout_v2.rb

```diff
@@ -459,37 +452,48 @@ def access_token_url
       end
 
       def setup_access_token
-        request = 'grant_type=client_credentials'
-        begin
-          raw_response = ssl_post(access_token_url, request, access_token_header)
-        rescue ResponseError => e
-          raise OAuthResponseError.new(e)
-        else
-          response = parse(raw_response)
+        response = parse(ssl_post(access_token_url, 'grant_type=client_credentials', access_token_header))
+        @options[:access_token] = response['access_token']
+        @options[:expires] = response['expires_in'] && response['expires_in'] > 0 && (DateTime.now + (response['expires_in'] - 120).seconds).strftime('%Q').to_i
 
-          if (access_token = response['access_token'])
-            access_token
-          else
-            raise OAuthResponseError.new(response)
-          end
-        end
+        Response.new(
+          response['access_token'].present?,
+          message_from(true, response, {}),
+          response.merge({ expires: @options[:expires] }),
+          test: test?,
+          error_code: error_code_from(false, response, {})
+        )
+      rescue ResponseError => e
+        raise OAuthResponseError.new(e)
       end
 
-      def commit(action, post, options, authorization = nil, method = :post)
+      def access_token_valid?
+        @options[:access_token].present? && @options[:expires] && @options[:expires].to_i > DateTime.now.strftime('%Q').to_i
+      end
+
+      def peform_payin(action, post, options, authorization = nil, method = :post)
         begin
           raw_response = ssl_request(method, url(action, authorization), post.nil? || post.empty? ? nil : post.to_json, headers(action, options))
           response = parse(raw_response)
           response['id'] = response['_links']['payment']['href'].split('/')[-1] if action == :capture && response.key?('_links')
-          source_id = authorization if action == :unstore
         rescue ResponseError => e
+          @options[:access_token] = '' if e.response.code == '401'
```

Comment:
I would suggest just to added check here for access_token in case the the transaction fails when using the secret_key

Reply:
good catch!

Reply:
@almalee24 I'm adding a test for this but it doesn't make sense since the `access_token` takes precedence over `secret_key` according to this [line of code](https://github.com/activemerchant/active_merchant/pull/5098/files#diff-ac1e1ae60a6cc04111367e448e705b6e74104fd908e64d1ffe6c7237b5deb65dR515)

Reply:
Yeah I guess it's not a big issue but if secret_key transaction fails with 401 then we do `@options[:access_token] = ''` which we don't need to do because we didn't use access_token. It's not breaking anything so it's not something you need to change but a nice improvement I guess.